src/app/api/toy-example/README.md

```diff
@@ -0,0 +1,398 @@
+# Developing RESTful API
+
+Hello, aspiring backend developers. In this tutorial, I will walk you through the design process for a RESTful API route.
+Do note that there are multiple "routes" you can take to design an effective API. The approach that I will be showing you
+is highly opinionated.
+
+## A Gentle Introduction
+
+RESTful API is a paradigm for client-server interaction using HTTP requests. The client makes a network request to the
+server with specified urls, such as `localhost:3000/api/login` and the server performs an action. The client can specify what action to perform by specifying a [HTTP verb](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods). For the scope of this tutorial, I will discuss about RESTful API as a paradigm to perform
+[CRUD operations](https://www.crowdstrike.com/cybersecurity-101/observability/crud/), meaning that it is used to interact with records in our database. The convention is as follow:
+
+- `GET` retrieve record(s) from the database
+- `POST` adds record(s) to the database
+- `PUT` updates existing record(s) in the database
+- `DELETE` removes record(s) from the database
+
+Suppose I want to design an API for professors to manage students' information. I may want to implement the following routes:
+
+1. `GET /api/students` to retrieve a list of every students that I have,
+2. `GET /api/students/[id]` to retrieve information for a specific student with the given `id`,
+3. `POST /api/students` to add a new student to the database,
+4. `PATCH /api/students/[id]` to update information for a specific student with the given `id`.
+
+As the developer of the API routes, you get to make the contract between the client and the server. In other words, you're telling the client "when url `X` is entered with HTTP verb `Y`, the server will perform action `Z`". Therefore, it's equally valid to design the abovementioned example with:
+
+1. `GET /api/getAllStudents`,
+2. `GET /api/getStudentByID`,
+3. `POST /api/createStudent`,
+4. `PATCH /api/updateStudentByID`.
+
+I opt-in for the first approach whenever possible; the segments of the API route specifies the collections that they interact with and the HTTP verb dictates what action to perform on these collections. I think this approach is more maintable and intuitive to end-users.
```

Comment:
Not really intuitive to end-users but intuitive for _other programmers_ on your team (or your future self) (which I'm sure is what you meant but maybe worth clarifying)